design/resource-filter-field-selectors-design.md

```diff
@@ -0,0 +1,158 @@
+# Extend Resource Policy To Support Field Selectors
+
+## Abstract
+
+This design proposal extends the [`ResourcePolicies` API][0] to support resource filtering using [field selectors][1].
+The proposed field selection design complements the existing [filtering API][2], to enable users to filter backup resources by Kubernetes resource fields.
+
+## Background
+
+The current mechanism to include cluster-scoped resources in backups via the `IncludeResources`, `IncludeClusterResources` and `IncludeClusterScopedResources` API filters the resources by their [`ResourceGroup`][3].
+This is insufficent for operator backup use cases, where specific Custom Resource Definition (CRD) must be included in order for the restored operator to work.
+See issue [#4876][4].
+
+The existing [label selection][5] mechanism doesn't scale on large Kubernetes clusters, due to constraints beyond the users' control like:
+
+* Inconsistent resource labeling scheme across different providers
+* Unique labeling to correctly group CRDs according to operators
+* Strict cluster-scoped RBAC permissions (to label CRDs)
+* Dynamically generated custom resource definitions
+
+## Goals
+
+- Users can define backup filter criteria based on Kubernetes field selectors.
+
+## Non Goals
+
+- N/A
+
+## High-Level Design
+
+Extends the `ResourcePolicies` API with a `FieldSelectorPolicies` configuration.
+
+## Detailed Design
+
+The `FieldSelectorPolicies` works in complement to the `IncludeClusterScopedResources` API.
+For example, to include the Nginx ingress controller CRD, one can define a field selector policy that looks like this:
+
+```sh
+cat <<EOF > backup-policy-nginx-crd.yaml
+version: v1
+fieldSelectorPolicies:
+- groupResource: 
+    group: apiextensions.k8s.io
+    resource: customresourcedefinitions
+  metadata.name: virtualservers.k8s.nginx.org
+  metadata.name: virtualserverroutes.k8s.nginx.org
+  metadata.name: transportservers.k8s.nginx.org
+  metadata.name: policies.k8s.nginx.org
+EOF
+```
+
+This YAML policy is used to create the policy configmap used by Velero to backup the CRDs:
+
+```sh
+kubectl -n velero create cm backup-policy-nginx-crd --from-file backup-policy-nginx-crd.yaml
+
+velero backup create --include-cluster-scoped-resources="customresourcedefinitions" --resource-policies-configmap=backup-policy-nginx-crd
+```
+
+Although the supported field selectors vary across Kubernetes resource types, all resource types support the `metadata.name` and `metadata.namespace` fields.
```

Comment:
Are we suggesting that we only allow these two in our API? 

If we allow others that a given resource type supports if there is an error, do we error the backup at that point? Is there someway to validate the request during backup resource creation vs during runtime?

Reply:
The goal is to let K8s tells us which field selectors are supported for a particular resource group, by e.g., specifying the field selectors in the [`metav1.ListOptions`](https://pkg.go.dev/k8s.io/apimachinery@v0.26.1/pkg/apis/meta/v1#ListOptions). Velero shouldn't need to maintain any knowledge of which field selectors are supported server-side.

As for error handling, not sure if it's worth doing more upfront client-side validation, since K8s already has a well-defined API for field selections. I think we can just let Kubernetes handle it, like how `kubectl get` works. If we got an error from K8s, Velero should fail the backup.

Reply:
Or some form of dry run option for the overall resource policy may help too? Personally, I am not a fan of admission webhooks because they can be disabled from the webhook configuration, they can go down, they don't load-balance correctly in HA etc.

Reply:
dry run to validate valid spec sounds good to me.

Reply:
I think that a dry-run command could take the CR and validate the FieldSelectors against a given cluster, would make sense as an option, and would give you a similar feel to the `kubectl get`.

At some point, I do wonder if the project is going to have to consider a webhook, but for this, I think the proposed idea of a dry run makes sense and would alleviate my concerns